encoders/protoc-gen-firebase-encoders/src/main/kotlin/com/google/firebase/encoders/proto/codegen/Parsing.kt

```diff
@@ -27,9 +33,128 @@ interface DescriptorParser {
+ *
+ * Any extensions that are encountered are added directly to extended messages as fields, so
+ * extensions are indistinguishable from message's own fields.
+ */
 class DefaultParser @Inject constructor(private val config: CodeGenConfig) : DescriptorParser {
-    override fun parse(protoFiles: List<FileDescriptorProto>): Collection<UserDefined> {
-        return listOf()
+    override fun parse(files: List<FileDescriptorProto>): List<UserDefined> {
+        val parsedMessages = files.asSequence().flatMap { file ->
+            val javaPackage = "${config.vendorPackage}.${file.`package`}"
+
+            val parent = Owner.Package(file.`package`, javaPackage, file.name)
+            parseEnums(parent, file.enumTypeList).plus(parseMessages(parent, file.messageTypeList))
+        }.toList()
+
+        val extensions = discoverExtensions(files)
+
+        val messagesWithExtensions = parsedMessages.map { msg ->
+            val msgExtensions = extensions[msg.protobufFullName]
+            if (msg !is Message || msgExtensions == null) {
+                return@map msg
+            }
+            msg.withMoreFields(msgExtensions)
+        }
+        resolveReferences(messagesWithExtensions)
+        return messagesWithExtensions.filter {
+            config.includeList.contains(it.protobufFullName)
+        }
+    }
+
+    private fun resolveReferences(types: Collection<UserDefined>) {
+        val messageIndex = types.asSequence()
+                .map { it.protobufFullName to it }
+                .toMap()
+        for (userDefined in types) {
+            if (userDefined !is Message) {
+                continue
+            }
+            for (field in userDefined.fields) {
+                (field.type as? Unresolved)?.run {
+                    field.type = messageIndex[this.protobufName]
+                                ?: throw IllegalArgumentException("Unresolved reference to $protobufName in ${userDefined.protobufFullName}.")
+                }
+            }
+        }
+    }
+
+    private fun discoverExtensions(
+        files: List<FileDescriptorProto>
+    ): ImmutableMultimap<String, ProtoField> {
+
+        val extensions: ImmutableListMultimap.Builder<String, ProtoField> =
+                ImmutableListMultimap.builder<String, ProtoField>()
+        for (file in files) {
+            for (field in file.extensionList) {
+                extensions.put(
+                    field.extendee.trimStart('.'),
+                    ProtoField(
+                        field.name,
+                        field.determineType(),
+                        field.number,
+                        field.label == FieldDescriptorProto.Label.LABEL_REPEATED))
+            }
+        }
+        return extensions.build()
+    }
+
+    private fun parseEnums(parent: Owner, enums: List<DescriptorProtos.EnumDescriptorProto>): Sequence<ProtoEnum> {
+        return enums.asSequence().map { enum ->
```

Comment:
curiosity, why don't you need to escape enum as \`enum\`?

Reply:
apparently Kotlin calls it a [Modifier keyword](https://kotlinlang.org/docs/reference/keyword-reference.html#modifier-keywords) which can be used as an identifier as well.